Update to jest 15

vjeux opened the pull request:
Jest 15 has just been released and is way more awesome: http://facebook.github.io/jest/blog/2016/09/01/jest-15.html

This pull request updates jest from version 12 to 15. Right now there's a fix in jest around Symbol ( https://github.com/facebook/jest/pull/1543 ) that hasn't been released yet, so this will break CI and cannot be merged it. But once it ships by the end of the day, we'll be good to go :)

See comments inline for all the changes I've done.

vjeux commented on package.json:
```diff
@@ -90,25 +90,19 @@
       "/.module-cache/",
       "/react/build/"
     ],
-    "persistModuleRegistryBetweenSpecs": true,
```
Jest upgrade path was so good. For each option that was deprecated, there was a big blob of text saying why and what was the new one I needed to use.

vjeux commented on package.json:
```diff
@@ -90,25 +90,19 @@
-      "ts"
-    ],
+    "testRegex": "/__tests__/",
+    "moduleFileExtensions": ["js", "json", "node", "coffee", "ts"],
```
Note that js, json and node are the default ones (that are needed to run the tests) and we need in addition coffee and ts for our integration testing with those.

zpao replied:
Actually, this will get reprinted too (I think it's just`JSON.stringify(data, null, 2)`). Let's change this and the one above to have newlines, like it currently is.

vjeux commented on src/isomorphic/modern/class/__tests__/ReactClassEquivalence-test.js:
```diff
@@ -32,7 +32,7 @@ describe('ReactClassEquivalence', function() {
   var cwd = process.cwd();
   var jestBin = path.resolve('node_modules', '.bin', 'jest');
-  var setupFile = path.resolve(__dirname, 'setupSpecEquivalenceReporter.js');
+  var setupFile = path.resolve(__dirname, 'setupSpecEquivalenceReporter.js.setup');
```
Jest now warns when it runs files that do not have any tests. We shouldn't even run this file through jest but because it is in the __tests__ folder and ends with .js, it is automatically being ran.

Adding .js.setup extension is a quick fix to make it work.

zpao replied:
What's the non-quick fix? Can we just do that? (I know how this works and we'll never come back to that)

vjeux replied:
We can either
- move the `setupSpecEquivalenceReporter.js` file outside of `__tests__` folder or
- hardcode that file name in the package.json config

I think the current solution is better than those two others. Let me know if you have a better idea

zpao replied:
What about putting it where we have other jasmine-related setups: root/scripts/jest? We're already making the assumption that `path.resolve` is going to work from our repo root so it just means we change `__dirname` to `scripts/jest`).

vjeux commented on src/isomorphic/modern/class/__tests__/ReactClassEquivalence-test.js:
```diff
@@ -60,7 +60,7 @@ function runJest(testFile) {
 
 function compareResults(a, b) {
-  var regexp = /^EQUIVALENCE.*$/gm;
+  var regexp = /^[ ]+EQUIVALENCE.*$/gm;
```
Jest changed the way console.log are outputted by adding two spaces before

zpao replied:
`\s`? Or even just a space character without the brackets.

vjeux replied:
Will update to two spaces.

vjeux commented on src/isomorphic/modern/types/__tests__/ReactFlowPropTypes-test.js:
```diff
@@ -1,18 +0,0 @@
-
-describe('ReactFlowPropTypes', function() {
-
-  // TODO: Test Flow integration and ensure that prop types works.
```
Jest now marks test files with no assertions as failing. I also really dislike having dead code in the codebase. If we are going to test proptypes with flow one day, we'll recreate this file.

vjeux commented on src/renderers/art/__tests__/ReactART-test.js:
```diff
@@ -13,9 +13,6 @@
 'use strict';
 
-jest
-  .unmock('ReactART');
```
We manually call unmock on a bunch of file even though automocking is off for the repo. There's now a warning in Jest for that :)

zpao replied:
FWIW These files were copied directly from another repo (where they are mostly copied from www). I don't think we can remove that line internally so we'll have to have it in some copies.

It's ok to diverge a tiny bit in this part but if we ever make changes to the actual tests, let's make sure those propagate.

cpojer replied:
art is unmocked by default on www, so this should be fine.

vjeux commented on src/renderers/dom/server/__tests__/ReactServerRendering-test.js:
```diff
@@ -45,33 +45,33 @@ describe('ReactServerRendering', function() {
         <span>hello world</span>
       );
-      expect(response).toMatch(
+      expect(response).toMatch(new RegExp(
```
`toMatch` used to silently convert the string into a RegExp and no longer does it. Replacing the few call sites we had to match the new behavior. It's IMO a good thing to have to pass a RegExp to toMatch :)

zpao commented on package.json:
```diff
@@ -90,25 +90,19 @@
     "rootDir": "",
     "scriptPreprocessor": "scripts/jest/preprocessor.js",
-    "setupEnvScriptFile": "scripts/jest/environment.js",
+    "setupFiles": [ "scripts/jest/environment.js" ],
```
Can you lose the spaces? They'll get wiped away when something like `npm install -S` happens anyway so let's avoid the future noise.

keyanzhang replied:
nit: no spaces in brackets?

vjeux replied:
Just did `npm install --save-dev some-random-module` and reprinted it correctly. I had no idea this was a thing, TIL :)

vjeux commented:
- Removed `^` from regex
- Reprinted package.json

cpojer commented:
Ok, 15.0.1 with a fix for the Symbol issue is out.

cpojer commented:
oops! wrong button.

vjeux commented:
- Updated to jest 15.0.1
- Moved the setup file in scripts/jest

This should be ready to go if someone wants to do a final review and do a stamp of approval :)

brigand commented:
You didn't change the gh review label btw.

cpojer commented:
woops! Sorry about that.
